### PR TITLE
[WIP] New query parameter for inclusion additional qualifiers

### PR DESCRIPTION
**Pull request by timgerstel**
Added new request parameter indicating whether or not to include additional qualifiers for dataset searches.  Modified query input to more closely represent 3.4 searching

Do not merge without:
https://github.com/zowe/zss/pull/75

Related to:
https://github.com/zowe/zlux-platform/pull/34
https://github.com/zowe/zlux-app-manager/pull/138
https://github.com/zowe/zlux-file-explorer/pull/33

Signed-off-by: Timothy Gerstel <tim.gerstel@gmail.com>

### CODE REVIEW

**Comment by timgerstel** [comment]
@jordanfilteau1995 I cant add you as a reviewer

File: c/datasetjson.c [hunk 1]
```diff
@@ -1756,7 +1759,27 @@ void respondWithDatasetMetadata(HttpResponse *response) {
 
   int fieldCount = defaultCSIFieldCount;
   char **csiFields = defaultCSIFields;
-
+  int asteriskPos = indexOf(dsn,dsnLen,'*',0);
+  int dblAsteriskPos = indexOfString(dsn,dsnLen,"**",0);
+  int periodPos = lastIndexOf(dsn, dsnLen, '.');
+  int addQualifiers = !strcmp(addQualifiersArg, "true");
+  
+  if(addQualifiers){
```
Comment (jordanfilteau1995) [inline c/datasetjson.c]: Please fix the white space problems. Check that you have spaces instead of tabs and that the indent is set to 2 instead of 4.

File: c/datasetjson.c [hunk 2]
```diff
@@ -1756,7 +1759,27 @@ void respondWithDatasetMetadata(HttpResponse *response) {
...
+	  int newDsnLen = dsnLen + 4; //+2 for asterisk, +1 for period, +1 for null terminator
+	  char newDsn[newDsnLen];
```
Comment (jordanfilteau1995) [inline c/datasetjson.c]: Try to stay away from variable length arrays. Use a define instead as the size of the array. Please check that your newDsnLen will not go over the value that you define to prevent copying out of bounds.

File: c/datasetjson.c [hunk 3]
```diff
@@ -1756,7 +1759,27 @@ void respondWithDatasetMetadata(HttpResponse *response) {
...
+	  strncpy(newDsn, dsn, dsnLen + 1);
+	  if(asteriskPos < 0 && dblAsteriskPos < 0){
+		  if(periodPos < 0 || periodPos != dsnLen - 1){
+			strncat(newDsn, ".**\0", 4);
+		  } else if(periodPos == dsnLen - 1){
+			  strncat(newDsn, "**\0", 3);
+		  }
+	  } else if(periodPos < 0 && asteriskPos == dsnLen - 1){
```
Comment (jordanfilteau1995) [inline c/datasetjson.c]: Like we talked about, I don't think the periodPos is necessary here.

File: c/datasetjson.c [hunk 4]
```diff
@@ -1756,7 +1759,27 @@ void respondWithDatasetMetadata(HttpResponse *response) {
...
-
+  int asteriskPos = indexOf(dsn,dsnLen,'*',0);
+  int dblAsteriskPos = indexOfString(dsn,dsnLen,"**",0);
+  int periodPos = lastIndexOf(dsn, dsnLen, '.');
+  int addQualifiers = !strcmp(addQualifiersArg, "true");
+  
+  if(addQualifiers){
+	  int newDsnLen = dsnLen + 4; //+2 for asterisk, +1 for period, +1 for null terminator
+	  char newDsn[newDsnLen];
+	  strncpy(newDsn, dsn, dsnLen + 1);
```
Comment (jordanfilteau1995) [inline c/datasetjson.c]: Doing a snprintf might be more elegant than doing a strncpy and then concatenating.

File: c/datasetjson.c [hunk 5]
```diff
@@ -1756,7 +1759,27 @@ void respondWithDatasetMetadata(HttpResponse *response) {
 
   int fieldCount = defaultCSIFieldCount;
   char **csiFields = defaultCSIFields;
-
+  int asteriskPos = indexOf(dsn,dsnLen,'*',0);
```
Comment (jordanfilteau1995) [inline c/datasetjson.c]: You might want to check if addQualifiers is on before you start getting positions in the string. You may be doing work for no reason. You can do all of these calculations inside the if statement. I don't see you using them anywhere else in the code.

**Comment by timgerstel** [comment]
Don't merge this yet, I still need to test edge cases.  Just committing so I don't lose work.
